diff --git a/dist/index.d.ts b/dist/index.d.ts
index d7adc46c0aafb8fd347c9a188e27430ca800c70f..043d506441b653481516284c29c9594796db7fcd 100644
--- a/dist/index.d.ts
+++ b/dist/index.d.ts
@@ -880,7 +880,7 @@ declare class TextInputBuilder extends ComponentBuilder<APITextInputComponent> i
 declare type MessageComponentBuilder = ActionRowBuilder<MessageActionRowComponentBuilder> | MessageActionRowComponentBuilder;
 declare type ModalComponentBuilder = ActionRowBuilder<ModalActionRowComponentBuilder> | ModalActionRowComponentBuilder;
 declare type MessageActionRowComponentBuilder = ButtonBuilder | SelectMenuBuilder;
-declare type ModalActionRowComponentBuilder = TextInputBuilder;
+declare type ModalActionRowComponentBuilder = TextInputBuilder | SelectMenuBuilder;
 declare type AnyComponentBuilder = MessageActionRowComponentBuilder | ModalActionRowComponentBuilder;
 /**
  * Represents an action row component
@@ -1072,7 +1072,7 @@ declare class SlashCommandBooleanOption extends ApplicationCommandOptionBase {
     toJSON(): APIApplicationCommandBooleanOption;
 }
 
-declare const allowedChannelTypes: readonly [ChannelType.GuildText, ChannelType.GuildVoice, ChannelType.GuildCategory, ChannelType.GuildNews, ChannelType.GuildNewsThread, ChannelType.GuildPublicThread, ChannelType.GuildPrivateThread, ChannelType.GuildStageVoice];
+declare const allowedChannelTypes: readonly [ChannelType.GuildText, ChannelType.GuildVoice, ChannelType.GuildCategory, ChannelType.GuildAnnouncement, ChannelType.AnnouncementThread, ChannelType.PublicThread, ChannelType.PrivateThread, ChannelType.GuildStageVoice];
 declare type ApplicationCommandOptionAllowedChannelTypes = typeof allowedChannelTypes[number];
 declare class ApplicationCommandOptionChannelTypesMixin {
     readonly channel_types?: ApplicationCommandOptionAllowedChannelTypes[];
diff --git a/dist/index.js b/dist/index.js
index 95f933e601691026107ab8493ba783e650754c68..87ea78e23c0c88c86912e041f848211450966617 100644
--- a/dist/index.js
+++ b/dist/index.js
@@ -998,10 +998,10 @@ var allowedChannelTypes = [
   import_v1011.ChannelType.GuildText,
   import_v1011.ChannelType.GuildVoice,
   import_v1011.ChannelType.GuildCategory,
-  import_v1011.ChannelType.GuildNews,
-  import_v1011.ChannelType.GuildNewsThread,
-  import_v1011.ChannelType.GuildPublicThread,
-  import_v1011.ChannelType.GuildPrivateThread,
+  import_v1011.ChannelType.GuildAnnouncement,
+  import_v1011.ChannelType.AnnouncementThread,
+  import_v1011.ChannelType.PublicThread,
+  import_v1011.ChannelType.PrivateThread,
   import_v1011.ChannelType.GuildStageVoice
 ];
 var channelTypesPredicate = import_shapeshift6.s.array(import_shapeshift6.s.union(...allowedChannelTypes.map((type) => import_shapeshift6.s.literal(type))));
diff --git a/dist/index.mjs b/dist/index.mjs
index 2ac0e7de449181d36074cbe39e012aec2a1177d5..607dd28e0de378040a9f265874e12d3e32c8f3e6 100644
--- a/dist/index.mjs
+++ b/dist/index.mjs
@@ -919,10 +919,10 @@ var allowedChannelTypes = [
   ChannelType.GuildText,
   ChannelType.GuildVoice,
   ChannelType.GuildCategory,
-  ChannelType.GuildNews,
-  ChannelType.GuildNewsThread,
-  ChannelType.GuildPublicThread,
-  ChannelType.GuildPrivateThread,
+  ChannelType.GuildAnnouncement,
+  ChannelType.AnnouncementThread,
+  ChannelType.PublicThread,
+  ChannelType.PrivateThread,
   ChannelType.GuildStageVoice
 ];
 var channelTypesPredicate = s6.array(s6.union(...allowedChannelTypes.map((type) => s6.literal(type))));
