diff --git a/dist/index.d.ts b/dist/index.d.ts
index 7fbce0c0b5e4b62ace0c961370d034d4fc345b8c..9b1b2b8fd41cf7c472b6d69ef72b975b7bf9ffa4 100644
--- a/dist/index.d.ts
+++ b/dist/index.d.ts
@@ -571,7 +571,7 @@ declare abstract class ComponentBuilder<DataType extends Partial<APIBaseComponen
 declare type MessageComponentBuilder = MessageActionRowComponentBuilder | ActionRowBuilder<MessageActionRowComponentBuilder>;
 declare type ModalComponentBuilder = ModalActionRowComponentBuilder | ActionRowBuilder<ModalActionRowComponentBuilder>;
 declare type MessageActionRowComponentBuilder = ButtonBuilder | SelectMenuBuilder;
-declare type ModalActionRowComponentBuilder = TextInputBuilder;
+declare type ModalActionRowComponentBuilder = TextInputBuilder | SelectMenuBuilder;
 declare type AnyComponentBuilder = MessageActionRowComponentBuilder | ModalActionRowComponentBuilder;
 /**
  * Represents an action row component
