diff --git a/dist/index.d.ts b/dist/index.d.ts
index c8b0571b92fd34fc35e06f841a2791b17d34a1db..cfd54ea774fa9b4ef8dd53bcde5cbbff56f84b99 100644
--- a/dist/index.d.ts
+++ b/dist/index.d.ts
@@ -352,13 +352,13 @@ declare function time<S extends TimestampStylesString>(date: Date, style: S): `<
 /**
  * Formats the given timestamp into a short date-time string
  *
- * @param seconds - The time to format, represents an UNIX timestamp in seconds
+ * @param seconds - The time to format, represents an UNIX timestamp in milliseconds
  */
 declare function time<C extends number>(seconds: C): `<t:${C}>`;
 /**
  * Formats the given timestamp into a short date-time string
  *
- * @param seconds - The time to format, represents an UNIX timestamp in seconds
+ * @param seconds - The time to format, represents an UNIX timestamp in milliseconds
  * @param style - The style to use
  */
 declare function time<C extends number, S extends TimestampStylesString>(seconds: C, style: S): `<t:${C}:${S}>`;
@@ -730,7 +730,7 @@ declare class TextInputBuilder extends ComponentBuilder<APITextInputComponent> {
 declare type MessageComponentBuilder = MessageActionRowComponentBuilder | ActionRowBuilder<MessageActionRowComponentBuilder>;
 declare type ModalComponentBuilder = ModalActionRowComponentBuilder | ActionRowBuilder<ModalActionRowComponentBuilder>;
 declare type MessageActionRowComponentBuilder = ButtonBuilder | SelectMenuBuilder;
-declare type ModalActionRowComponentBuilder = TextInputBuilder;
+declare type ModalActionRowComponentBuilder = TextInputBuilder | SelectMenuBuilder;
 declare type AnyComponentBuilder = MessageActionRowComponentBuilder | ModalActionRowComponentBuilder;
 /**
  * Represents an action row component
diff --git a/dist/index.js b/dist/index.js
index 11dc178316016c109eddb33038e1bb0d16db6278..6500626752f354cdd39e379324712b16dc955919 100644
--- a/dist/index.js
+++ b/dist/index.js
@@ -337,9 +337,9 @@ function formatEmoji(emojiId, animated = false) {
 __name(formatEmoji, "formatEmoji");
 function time(timeOrSeconds, style) {
   if (typeof timeOrSeconds !== "number") {
-    timeOrSeconds = Math.floor((timeOrSeconds?.getTime() ?? Date.now()) / 1e3);
+    timeOrSeconds = timeOrSeconds?.getTime() ?? Date.now();
   }
-  return typeof style === "string" ? `<t:${timeOrSeconds}:${style}>` : `<t:${timeOrSeconds}>`;
+  return typeof style === "string" ? `<t:${Math.floor(timeOrSeconds / 1e3)}:${style}>` : `<t:${Math.floor(timeOrSeconds / 1e3)}>`;
 }
 __name(time, "time");
 var TimestampStyles = {
diff --git a/dist/index.mjs b/dist/index.mjs
index bce0a1b9b987cc1bf4a6b087b1c017bda4b06526..572b54bee3e07f3c5a7a6cfb8ec85de7ef89a46e 100644
--- a/dist/index.mjs
+++ b/dist/index.mjs
@@ -257,9 +257,9 @@ function formatEmoji(emojiId, animated = false) {
 __name(formatEmoji, "formatEmoji");
 function time(timeOrSeconds, style) {
   if (typeof timeOrSeconds !== "number") {
-    timeOrSeconds = Math.floor((timeOrSeconds?.getTime() ?? Date.now()) / 1e3);
+    timeOrSeconds = timeOrSeconds?.getTime() ?? Date.now();
   }
-  return typeof style === "string" ? `<t:${timeOrSeconds}:${style}>` : `<t:${timeOrSeconds}>`;
+  return typeof style === "string" ? `<t:${Math.floor(timeOrSeconds / 1e3)}:${style}>` : `<t:${Math.floor(timeOrSeconds / 1e3)}>`;
 }
 __name(time, "time");
 var TimestampStyles = {
