diff --git a/dist/components/ActionRow.d.ts b/dist/components/ActionRow.d.ts
index 9b03460050f40dce964f589fa70d4c1a2cd9d67b..4983bdf3f9708a453e3214c61b8135acd0dfa201 100644
--- a/dist/components/ActionRow.d.ts
+++ b/dist/components/ActionRow.d.ts
@@ -7,7 +7,7 @@ import type { TextInputBuilder } from './textInput/TextInput.js';
 export declare type MessageComponentBuilder = ActionRowBuilder<MessageActionRowComponentBuilder> | MessageActionRowComponentBuilder;
 export declare type ModalComponentBuilder = ActionRowBuilder<ModalActionRowComponentBuilder> | ModalActionRowComponentBuilder;
 export declare type MessageActionRowComponentBuilder = ButtonBuilder | SelectMenuBuilder;
-export declare type ModalActionRowComponentBuilder = TextInputBuilder;
+export declare type ModalActionRowComponentBuilder = TextInputBuilder | SelectMenuBuilder;
 export declare type AnyComponentBuilder = MessageActionRowComponentBuilder | ModalActionRowComponentBuilder;
 /**
  * Represents an action row component
diff --git a/dist/interactions/slashCommands/mixins/ApplicationCommandOptionChannelTypesMixin.cjs b/dist/interactions/slashCommands/mixins/ApplicationCommandOptionChannelTypesMixin.cjs
index ae13bfbd544d42fbdca78dbc88df22670be3142d..36a19a9555bca5a1ecbbcbf58e6f81563c0dfbb8 100644
--- a/dist/interactions/slashCommands/mixins/ApplicationCommandOptionChannelTypesMixin.cjs
+++ b/dist/interactions/slashCommands/mixins/ApplicationCommandOptionChannelTypesMixin.cjs
@@ -9,10 +9,10 @@ const allowedChannelTypes = [
   v10.ChannelType.GuildText,
   v10.ChannelType.GuildVoice,
   v10.ChannelType.GuildCategory,
-  v10.ChannelType.GuildNews,
-  v10.ChannelType.GuildNewsThread,
-  v10.ChannelType.GuildPublicThread,
-  v10.ChannelType.GuildPrivateThread,
+  v10.ChannelType.GuildAnnouncement,
+  v10.ChannelType.AnnouncementThread,
+  v10.ChannelType.PublicThread,
+  v10.ChannelType.PrivateThread,
   v10.ChannelType.GuildStageVoice
 ];
 const channelTypesPredicate = shapeshift.s.array(shapeshift.s.union(...allowedChannelTypes.map((type) => shapeshift.s.literal(type))));
diff --git a/dist/interactions/slashCommands/mixins/ApplicationCommandOptionChannelTypesMixin.d.ts b/dist/interactions/slashCommands/mixins/ApplicationCommandOptionChannelTypesMixin.d.ts
index deb4fe1e846d1af9b5d419efbc08df57872e27c4..ed1944f7538baaf9f362b5666080ee52905b0843 100644
--- a/dist/interactions/slashCommands/mixins/ApplicationCommandOptionChannelTypesMixin.d.ts
+++ b/dist/interactions/slashCommands/mixins/ApplicationCommandOptionChannelTypesMixin.d.ts
@@ -1,5 +1,5 @@
 import { ChannelType } from 'discord-api-types/v10';
-declare const allowedChannelTypes: readonly [ChannelType.GuildText, ChannelType.GuildVoice, ChannelType.GuildCategory, ChannelType.GuildNews, ChannelType.GuildNewsThread, ChannelType.GuildPublicThread, ChannelType.GuildPrivateThread, ChannelType.GuildStageVoice];
+declare const allowedChannelTypes: readonly [ChannelType.GuildText, ChannelType.GuildVoice, ChannelType.GuildCategory, ChannelType.GuildAnnouncement, ChannelType.AnnouncementThread, ChannelType.PublicThread, ChannelType.PrivateThread, ChannelType.GuildStageVoice];
 export declare type ApplicationCommandOptionAllowedChannelTypes = typeof allowedChannelTypes[number];
 export declare class ApplicationCommandOptionChannelTypesMixin {
     readonly channel_types?: ApplicationCommandOptionAllowedChannelTypes[];
diff --git a/dist/interactions/slashCommands/mixins/ApplicationCommandOptionChannelTypesMixin.mjs b/dist/interactions/slashCommands/mixins/ApplicationCommandOptionChannelTypesMixin.mjs
index e2be25e02b29b1fa5e1484779231cd7d8aea1fa0..a4d4944117d170377eae3c64fecc0c2fdbb5cd45 100644
--- a/dist/interactions/slashCommands/mixins/ApplicationCommandOptionChannelTypesMixin.mjs
+++ b/dist/interactions/slashCommands/mixins/ApplicationCommandOptionChannelTypesMixin.mjs
@@ -5,10 +5,10 @@ const allowedChannelTypes = [
   ChannelType.GuildText,
   ChannelType.GuildVoice,
   ChannelType.GuildCategory,
-  ChannelType.GuildNews,
-  ChannelType.GuildNewsThread,
-  ChannelType.GuildPublicThread,
-  ChannelType.GuildPrivateThread,
+  ChannelType.GuildAnnouncement,
+  ChannelType.AnnouncementThread,
+  ChannelType.PublicThread,
+  ChannelType.PrivateThread,
   ChannelType.GuildStageVoice
 ];
 const channelTypesPredicate = s.array(s.union(...allowedChannelTypes.map((type) => s.literal(type))));
