diff --git a/lib/type.cc b/lib/type.cc
index dcb6176543056ad9be66ad4e3903a4b80fcc31b9..c638a744263262029d80ead3d0beaa7708a928d7 100644
--- a/lib/type.cc
+++ b/lib/type.cc
@@ -45,7 +45,7 @@ void GetProxyDetails(const Nan::FunctionCallbackInfo<v8::Value>& args) {
 }
 
 void Init(v8::Local<v8::Object> exports) {
-	v8::Local<v8::Context> context = exports->CreationContext();
+	v8::Local<v8::Context> context = exports->GetCreationContextChecked();
 
 	exports->Set(context, Nan::New("getPromiseDetails").ToLocalChecked(), Nan::New<v8::FunctionTemplate>(GetPromiseDetails)->GetFunction(context).ToLocalChecked());
 	exports->Set(context, Nan::New("getProxyDetails").ToLocalChecked(), Nan::New<v8::FunctionTemplate>(GetProxyDetails)->GetFunction(context).ToLocalChecked());
