diff --git a/dist/index.mjs b/dist/index.mjs
index 07bb51d6bdd68a7c4798be689d1b6b30831f6df6..25195cf37871e384fc9337909e3f88a83759eaa8 100644
--- a/dist/index.mjs
+++ b/dist/index.mjs
@@ -47,7 +47,7 @@ var ButtonLimits = {
 var SelectMenuLimits = {
   MaximumCustomIdCharacters: 100,
   MaximumOptionsLength: 25,
-  MaximumPlaceholderCharacters: 100,
+  MaximumPlaceholderCharacters: 150,
   MaximumMinValuesSize: 25,
   MaximumMaxValuesSize: 25,
   MaximumLengthOfNameOfOption: 100
@@ -73,7 +73,7 @@ var UserLimits = {
   MaximumBiographyLength: 190
 };
 var AutoCompleteLimits = {
-  MaximumAmountOfOptions: 20,
+  MaximumAmountOfOptions: 25,
   MaximumLengthOfNameOfOption: 100
 };
 var ModalLimits = {
