diff --git a/globals.d.ts b/globals.d.ts
index 08dc5269fb60a36f4aaab9a414fe1c51e6a5d525..196242591bb765b1f84eb64ae7038d878fe09864 100755
--- a/globals.d.ts
+++ b/globals.d.ts
@@ -53,7 +53,7 @@ interface AbortController {
     /**
      * Invoking this method will set this object's AbortSignal's aborted flag and signal to any observers that the associated activity is to be aborted.
      */
-    abort(): void;
+    abort(reason?: unknown): void;
 }
 
 /** A signal object that allows you to communicate with a DOM request (such as a Fetch) and abort it if required via an AbortController object. */
@@ -62,6 +62,16 @@ interface AbortSignal extends EventTarget {
      * Returns true if this AbortSignal's AbortController has signaled to abort, and false otherwise.
      */
     readonly aborted: boolean;
+    /**
+     * An optional reason specified when the AbortSignal was triggered.
+     * @since v17.2.0
+     */
+    readonly reason: unknown;
+    /**
+     * If {@link aborted} is true, throws {@link reason}.
+     * @since v17.3.0
+     */
+    throwIfAborted(): void;
 }
 
 declare var AbortController: {
