diff --git a/dist/chart.mjs b/dist/chart.mjs
index 6f506902d31401c7031ea75375dee146aa219495..52969872fcf918f2eee995ca3b5a1a0e82249ac1 100644
--- a/dist/chart.mjs
+++ b/dist/chart.mjs
@@ -5434,7 +5434,7 @@ class Chart {
     this.platform = new (config.platform || _detectPlatform(initialCanvas))();
     this.platform.updateConfig(config);
     const context = this.platform.acquireContext(initialCanvas, options.aspectRatio);
-    const canvas = context && context.canvas;
+    const canvas = context && context.canvas || initialCanvas;
     const height = canvas && canvas.height;
     const width = canvas && canvas.width;
     this.id = uid();
