diff --git a/dist/chart.js b/dist/chart.js
index eead2d66d781a129d532a99a0e395b36d36d22c2..99c25b9ad52875e9820d189a01611b05c34a4f45 100644
--- a/dist/chart.js
+++ b/dist/chart.js
@@ -5553,7 +5553,7 @@ class Chart {
         this.platform = new (config.platform || _detectPlatform(initialCanvas))();
         this.platform.updateConfig(config);
         const context = this.platform.acquireContext(initialCanvas, options.aspectRatio);
-        const canvas = context && context.canvas;
+        const canvas = context && context.canvas || initialCanvas;
         const height = canvas && canvas.height;
         const width = canvas && canvas.width;
         this.id = uid();
