diff --git a/src/client/websocket/WebSocketShard.js b/src/client/websocket/WebSocketShard.js
index 756cd08ef4cdd1e791487f83f6592d08f3d3c370..3f8541aa69cce977defd11e2d79188c0713052fd 100644
--- a/src/client/websocket/WebSocketShard.js
+++ b/src/client/websocket/WebSocketShard.js
@@ -269,6 +269,7 @@ class WebSocketShard extends EventEmitter {
       this.status = this.status === Status.Disconnected ? Status.Reconnecting : Status.Connecting;
       this.setHelloTimeout();
 
+      this.setWsCloseTimeout(-1);
       this.connectedAt = Date.now();
 
       // Adding a handshake timeout to just make sure no zombie connection appears.
@@ -368,7 +369,7 @@ class WebSocketShard extends EventEmitter {
     this.setWsCloseTimeout(-1);
     // If we still have a connection object, clean up its listeners
     if (this.connection) this._cleanupConnection();
-    this.status = Status.DISCONNECTED;
+    this.status = Status.Disconnected;
     this.emitClose(event);
   }
 
@@ -833,8 +834,8 @@ class WebSocketShard extends EventEmitter {
     // Step 2: Null the connection object
     this.connection = null;
 
-    // Step 3: Set the shard status to DISCONNECTED
-    this.status = Status.DISCONNECTED;
+    // Step 3: Set the shard status to Disconnected
+    this.status = Status.Disconnected;
 
     // Step 4: Cache the old sequence (use to attempt a resume)
     if (this.sequence !== -1) this.closeSequence = this.sequence;
diff --git a/src/structures/ActionRowBuilder.js b/src/structures/ActionRowBuilder.js
index 871d519533caa340099f898fc81da0efa56432da..bfa68e0d3b78d306a763ed030ad9fefa814f845e 100644
--- a/src/structures/ActionRowBuilder.js
+++ b/src/structures/ActionRowBuilder.js
@@ -9,13 +9,6 @@ const Transformers = require('../util/Transformers');
  * @extends {BuildersActionRow}
  */
 class ActionRowBuilder extends BuildersActionRow {
-  constructor({ components, ...data } = {}) {
-    super({
-      ...Transformers.toSnakeCase(data),
-      components: components?.map(c => (c instanceof ComponentBuilder ? c : Components.createComponentBuilder(c))),
-    });
-  }
-
   /**
    * Creates a new action row builder from JSON data
    * @param {JSONEncodable<APIActionRowComponent<APIActionRowComponentTypes>>
diff --git a/src/structures/AttachmentBuilder.js b/src/structures/AttachmentBuilder.js
index fe4ce59a2ad3be68cb3313314a08ae993de8c039..5e6adae3599626e49772ef356bfd06204f8b43e0 100644
--- a/src/structures/AttachmentBuilder.js
+++ b/src/structures/AttachmentBuilder.js
@@ -7,25 +7,19 @@ const Util = require('../util/Util');
  */
 class AttachmentBuilder {
   /**
-   * @param {BufferResolvable|Stream} attachment The file
-   * @param {APIAttachment} [data] Extra data
+   * @param {AttachmentPayload} payload Attachment and extra data
    */
-  constructor(attachment, data = {}) {
+  constructor({ attachment, ...data } = {}) {
     /**
      * The file associated with this attachment.
      * @type {BufferResolvable|Stream}
      */
     this.attachment = attachment;
     /**
-     * The name of this attachment
-     * @type {?string}
-     */
-    this.name = data.name;
-    /**
-     * The description of the attachment
-     * @type {?string}
+     * The API data associated with this attachment.
+     * @type {APIAttachment}
      */
-    this.description = data.description;
+    this.data = data;
   }
 
   /**
@@ -34,7 +28,7 @@ class AttachmentBuilder {
    * @returns {AttachmentBuilder} This attachment
    */
   setDescription(description) {
-    this.description = description;
+    this.data.description = description;
     return this;
   }
 
@@ -54,7 +48,7 @@ class AttachmentBuilder {
    * @returns {AttachmentBuilder} This attachment
    */
   setName(name) {
-    this.name = name;
+    this.data.name = name;
     return this;
   }
 
@@ -68,7 +62,7 @@ class AttachmentBuilder {
 
     if (!spoiler) {
       while (this.spoiler) {
-        this.name = this.name.slice('SPOILER_'.length);
+        this.data.name = this.data.name.slice('SPOILER_'.length);
       }
       return this;
     }
@@ -82,11 +76,11 @@ class AttachmentBuilder {
    * @readonly
    */
   get spoiler() {
-    return Util.basename(this.name).startsWith('SPOILER_');
+    return Util.basename(this.data.name ?? '').startsWith('SPOILER_');
   }
 
   toJSON() {
-    return Util.flatten(this);
+    return { ...this.data, attachment: this.attachment };
   }
 
   /**
@@ -95,10 +89,10 @@ class AttachmentBuilder {
    * @returns {AttachmentBuilder}
    */
   static from(other) {
-    return new AttachmentBuilder(other.attachment, {
-      name: other.name,
-      description: other.description,
-    });
+    if (isJSONEncodable(other)) {
+      return new this(other.toJSON());
+    }
+    return new this(other);
   }
 }
 
diff --git a/src/structures/ButtonBuilder.js b/src/structures/ButtonBuilder.js
index c427ce3c7516c88a36613255d6f7914aa7eb9ffa..0136a626eadbc74f61dec1853509cb3b63e8864b 100644
--- a/src/structures/ButtonBuilder.js
+++ b/src/structures/ButtonBuilder.js
@@ -2,31 +2,12 @@
 
 const { ButtonBuilder: BuildersButton, isJSONEncodable } = require('@discordjs/builders');
 const Transformers = require('../util/Transformers');
-const Util = require('../util/Util');
 
 /**
  * Represents a button builder.
  * @extends {BuildersButton}
  */
 class ButtonBuilder extends BuildersButton {
-  constructor({ emoji, ...data } = {}) {
-    super(
-      Transformers.toSnakeCase({ ...data, emoji: emoji && typeof emoji === 'string' ? Util.parseEmoji(emoji) : emoji }),
-    );
-  }
-
-  /**
-   * Sets the emoji to display on this button
-   * @param {string|APIMessageComponentEmoji} emoji The emoji to display on this button
-   * @returns {ButtonBuilder}
-   */
-  setEmoji(emoji) {
-    if (typeof emoji === 'string') {
-      return super.setEmoji(Util.parseEmoji(emoji));
-    }
-    return super.setEmoji(emoji);
-  }
-
   /**
    * Creates a new button builder from JSON data
    * @param {JSONEncodable<APIButtonComponent>|APIButtonComponent} other The other data
diff --git a/src/structures/GuildEmoji.js b/src/structures/GuildEmoji.js
index b0a8ff2fd421b6fee59a3376fd72485d2c23afda..ce920b9046f3677ffe23efb947313c2316f00cad 100644
--- a/src/structures/GuildEmoji.js
+++ b/src/structures/GuildEmoji.js
@@ -55,8 +55,9 @@ class GuildEmoji extends BaseGuildEmoji {
    * @readonly
    */
   get deletable() {
-    if (!this.guild.members.me) throw new Error('GUILD_UNCACHED_ME');
-    return !this.managed && this.guild.members.me.permissions.has(PermissionFlagsBits.ManageEmojisAndStickers);
+    const { me } = this.guild.members;
+    if (!me) throw new Error('GUILD_UNCACHED_ME');
+    return !this.managed && me.permissions.has(PermissionFlagsBits.ManageEmojisAndStickers);
   }
 
   /**
diff --git a/src/structures/GuildMember.js b/src/structures/GuildMember.js
index f732ca75400e7b799abd52ed401b08941f1a9251..a99cc2252dc3ea0de248eed185952dcac6a7e65c 100644
--- a/src/structures/GuildMember.js
+++ b/src/structures/GuildMember.js
@@ -239,8 +239,9 @@ class GuildMember extends Base {
     if (this.user.id === this.guild.ownerId) return false;
     if (this.user.id === this.client.user.id) return false;
     if (this.client.user.id === this.guild.ownerId) return true;
-    if (!this.guild.members.me) throw new Error('GUILD_UNCACHED_ME');
-    return this.guild.members.me.roles.highest.comparePositionTo(this.roles.highest) > 0;
+    const { me } = this.guild.members;
+    if (!me) throw new Error('GUILD_UNCACHED_ME');
+    return me.roles.highest.comparePositionTo(this.roles.highest) > 0;
   }
 
   /**
@@ -249,7 +250,6 @@ class GuildMember extends Base {
    * @readonly
    */
   get kickable() {
-    if (!this.guild.members.me) throw new Error('GUILD_UNCACHED_ME');
     return this.manageable && this.guild.members.me.permissions.has(PermissionFlagsBits.KickMembers);
   }
 
@@ -259,7 +259,6 @@ class GuildMember extends Base {
    * @readonly
    */
   get bannable() {
-    if (!this.guild.members.me) throw new Error('GUILD_UNCACHED_ME');
     return this.manageable && this.guild.members.me.permissions.has(PermissionFlagsBits.BanMembers);
   }
 
@@ -272,7 +271,7 @@ class GuildMember extends Base {
     return (
       !this.permissions.has(PermissionFlagsBits.Administrator) &&
       this.manageable &&
-      (this.guild.members.me?.permissions.has(PermissionFlagsBits.ModerateMembers) ?? false)
+      this.guild.members.me.permissions.has(PermissionFlagsBits.ModerateMembers)
     );
   }
 
diff --git a/src/structures/Invite.js b/src/structures/Invite.js
index 03b41500f925306a972fc3fc8972c7443e1bb654..2f184fd87582988bc97781643ec99c4a41a329c1 100644
--- a/src/structures/Invite.js
+++ b/src/structures/Invite.js
@@ -233,10 +233,11 @@ class Invite extends Base {
   get deletable() {
     const guild = this.guild;
     if (!guild || !this.client.guilds.cache.has(guild.id)) return false;
-    if (!guild.members.me) throw new Error('GUILD_UNCACHED_ME');
+    const { me } = guild.members;
+    if (!me) throw new Error('GUILD_UNCACHED_ME');
     return Boolean(
       this.channel?.permissionsFor(this.client.user).has(PermissionFlagsBits.ManageChannels, false) ||
-        guild.members.me.permissions.has(PermissionFlagsBits.ManageGuild),
+      me.permissions.has(PermissionFlagsBits.ManageGuild),
     );
   }
 
diff --git a/src/structures/Message.js b/src/structures/Message.js
index 32e6fd74e8e17dadf2f4d573bd98bf494042743c..e8f3b0688b3295ce03fc44e23d0b004c1ee4b28c 100644
--- a/src/structures/Message.js
+++ b/src/structures/Message.js
@@ -275,10 +275,13 @@ class Message extends Base {
       this.client.channels._add(data.thread, this.guild);
     }
 
-    if (this.member && data.member) {
-      this.member._patch(data.member);
-    } else if (data.member && this.guild && this.author) {
-      this.guild.members._add(Object.assign(data.member, { user: this.author }));
+    if (data.member) {
+      const { member } = this;
+      if (member) {
+        member._patch(data.member);
+      } else if (this.author) {
+        this.guild?.members._add(Object.assign(data.member, { user: this.author }));
+      }
     }
 
     if ('flags' in data) {
@@ -552,11 +555,16 @@ class Message extends Base {
    * @readonly
    */
   get editable() {
-    const precheck = Boolean(this.author.id === this.client.user.id && (!this.guild || this.channel?.viewable));
+    const { channel } = this;
+    const precheck = Boolean(
+      this.author.id === this.client.user.id &&
+      !this.flags.has(MessageFlags.Ephemeral) &&
+      (!this.guildId || channel?.viewable)
+    );
     // Regardless of permissions thread messages cannot be edited if
     // the thread is locked.
-    if (this.channel?.isThread()) {
-      return precheck && !this.channel.locked;
+    if (channel?.isThread()) {
+      return precheck && !channel.locked;
     }
     return precheck;
   }
@@ -567,23 +575,27 @@ class Message extends Base {
    * @readonly
    */
   get deletable() {
-    if (!this.guild) {
+    if (this.flags.has(MessageFlags.Ephemeral)) {
+      return false;
+    }
+    if (!this.guildId) {
       return this.author.id === this.client.user.id;
     }
     // DMChannel does not have viewable property, so check viewable after proved that message is on a guild.
-    if (!this.channel?.viewable) {
+    const { channel } = this;
+    if (!channel?.viewable) {
       return false;
     }
 
-    const permissions = this.channel?.permissionsFor(this.client.user);
+    const permissions = channel?.permissionsFor(this.client.user);
     if (!permissions) return false;
     // This flag allows deleting even if timed out
     if (permissions.has(PermissionFlagsBits.Administrator, false)) return true;
 
     return Boolean(
       this.author.id === this.client.user.id ||
-        (permissions.has(PermissionFlagsBits.ManageMessages, false) &&
-          this.guild.members.me.communicationDisabledUntilTimestamp < Date.now()),
+      (permissions.has(PermissionFlagsBits.ManageMessages, false) &&
+        this.guild.members.me.communicationDisabledUntilTimestamp < Date.now()),
     );
   }
 
@@ -596,9 +608,10 @@ class Message extends Base {
     const { channel } = this;
     return Boolean(
       !this.system &&
-        (!this.guild ||
-          (channel?.viewable &&
-            channel?.permissionsFor(this.client.user)?.has(PermissionFlagsBits.ManageMessages, false))),
+      !this.flags.has(MessageFlags.Ephemeral) &&
+      (!this.guildId ||
+        (channel?.viewable &&
+          channel?.permissionsFor(this.client.user)?.has(PermissionFlagsBits.ManageMessages, false))),
     );
   }
 
@@ -627,10 +640,10 @@ class Message extends Base {
     const { channel } = this;
     return Boolean(
       channel?.type === ChannelType.GuildNews &&
-        !this.flags.has(MessageFlags.Crossposted) &&
-        this.type === MessageType.Default &&
-        channel.viewable &&
-        channel.permissionsFor(this.client.user)?.has(bitfield, false),
+      !this.flags.any(MessageFlags.Crossposted | MessageFlags.Ephemeral) &&
+      this.type === MessageType.Default &&
+      channel.viewable &&
+      channel.permissionsFor(this.client.user)?.has(bitfield, false),
     );
   }
 
@@ -661,8 +674,9 @@ class Message extends Base {
    *   .catch(console.error);
    */
   edit(options) {
-    if (!this.channel) return Promise.reject(new Error('CHANNEL_NOT_CACHED'));
-    return this.channel.messages.edit(this, options);
+    const { channel } = this;
+    if (!channel) return Promise.reject(new Error('CHANNEL_NOT_CACHED'));
+    return channel.messages.edit(this, options);
   }
 
   /**
@@ -677,8 +691,9 @@ class Message extends Base {
    * }
    */
   crosspost() {
-    if (!this.channel) return Promise.reject(new Error('CHANNEL_NOT_CACHED'));
-    return this.channel.messages.crosspost(this.id);
+    const { channel } = this;
+    if (!channel) return Promise.reject(new Error('CHANNEL_NOT_CACHED'));
+    return channel.messages.crosspost(this.id);
   }
 
   /**
@@ -692,8 +707,9 @@ class Message extends Base {
    *   .catch(console.error)
    */
   async pin(reason) {
-    if (!this.channel) throw new Error('CHANNEL_NOT_CACHED');
-    await this.channel.messages.pin(this.id, reason);
+    const { channel } = this;
+    if (!channel) throw new Error('CHANNEL_NOT_CACHED');
+    await channel.messages.pin(this.id, reason);
     return this;
   }
 
@@ -708,8 +724,9 @@ class Message extends Base {
    *   .catch(console.error)
    */
   async unpin(reason) {
-    if (!this.channel) throw new Error('CHANNEL_NOT_CACHED');
-    await this.channel.messages.unpin(this.id, reason);
+    const { channel } = this;
+    if (!channel) throw new Error('CHANNEL_NOT_CACHED');
+    await channel.messages.unpin(this.id, reason);
     return this;
   }
 
@@ -729,13 +746,14 @@ class Message extends Base {
    *   .catch(console.error);
    */
   async react(emoji) {
-    if (!this.channel) throw new Error('CHANNEL_NOT_CACHED');
-    await this.channel.messages.react(this.id, emoji);
+    const { channel } = this;
+    if (!channel) throw new Error('CHANNEL_NOT_CACHED');
+    await channel.messages.react(this.id, emoji);
 
     return this.client.actions.MessageReactionAdd.handle(
       {
         user: this.client.user,
-        channel: this.channel,
+        channel,
         message: this,
         emoji: Util.resolvePartialEmoji(emoji),
       },
@@ -753,8 +771,9 @@ class Message extends Base {
    *   .catch(console.error);
    */
   async delete() {
-    if (!this.channel) throw new Error('CHANNEL_NOT_CACHED');
-    await this.channel.messages.delete(this.id);
+    const { channel } = this;
+    if (!channel) throw new Error('CHANNEL_NOT_CACHED');
+    await channel.messages.delete(this.id);
     return this;
   }
 
@@ -777,7 +796,8 @@ class Message extends Base {
    *   .catch(console.error);
    */
   reply(options) {
-    if (!this.channel) return Promise.reject(new Error('CHANNEL_NOT_CACHED'));
+    const { channel } = this;
+    if (!channel) return Promise.reject(new Error('CHANNEL_NOT_CACHED'));
     let data;
 
     if (options instanceof MessagePayload) {
@@ -790,7 +810,7 @@ class Message extends Base {
         },
       });
     }
-    return this.channel.send(data);
+    return channel.send(data);
   }
 
   /**
@@ -820,12 +840,13 @@ class Message extends Base {
    * @returns {Promise<ThreadChannel>}
    */
   startThread(options = {}) {
-    if (!this.channel) return Promise.reject(new Error('CHANNEL_NOT_CACHED'));
-    if (![ChannelType.GuildText, ChannelType.GuildNews].includes(this.channel.type)) {
+    const { channel } = this;
+    if (!channel) return Promise.reject(new Error('CHANNEL_NOT_CACHED'));
+    if (![ChannelType.GuildText, ChannelType.GuildNews].includes(channel.type)) {
       return Promise.reject(new Error('MESSAGE_THREAD_PARENT'));
     }
     if (this.hasThread) return Promise.reject(new Error('MESSAGE_EXISTING_THREAD'));
-    return this.channel.threads.create({ ...options, startMessage: this });
+    return channel.threads.create({ ...options, startMessage: this });
   }
 
   /**
@@ -834,8 +855,9 @@ class Message extends Base {
    * @returns {Promise<Message>}
    */
   fetch(force = true) {
-    if (!this.channel) return Promise.reject(new Error('CHANNEL_NOT_CACHED'));
-    return this.channel.messages.fetch({ message: this.id, force });
+    const { channel } = this;
+    if (!channel) return Promise.reject(new Error('CHANNEL_NOT_CACHED'));
+    return channel.messages.fetch({ message: this.id, force });
   }
 
   /**
diff --git a/src/structures/MessagePayload.js b/src/structures/MessagePayload.js
index 3f1769bcdb78873de1b37513e1bca33c4836b78a..e9bf087ca39bca2f9a78dbaa592c6e9f7e687322 100644
--- a/src/structures/MessagePayload.js
+++ b/src/structures/MessagePayload.js
@@ -105,7 +105,7 @@ class MessagePayload {
     if (this.options.content === null) {
       content = '';
     } else if (typeof this.options.content !== 'undefined') {
-      content = Util.verifyString(this.options.content, RangeError, 'MESSAGE_CONTENT_TYPE', true);
+      content = Util.verifyString(this.options.content, RangeError, 'MESSAGE_CONTENT_TYPE');
     }
 
     return content;
@@ -132,7 +132,7 @@ class MessagePayload {
       }
     }
 
-    const components = this.options.components?.map(c => (isJSONEncodable(c) ? c : new ActionRowBuilder(c)).toJSON());
+    const components = this.options.components?.map(c => isJSONEncodable(c) ? c.toJSON() : this.target.client.options.jsonTransformer(c));
 
     let username;
     let avatarURL;
diff --git a/src/structures/ModalSubmitFields.js b/src/structures/ModalSubmitFields.js
index 8929d54ed2f33e6409cf167215010c1548bef42a..ed3f1b8a300debb9dca2a37a227b65e89b0c86e6 100644
--- a/src/structures/ModalSubmitFields.js
+++ b/src/structures/ModalSubmitFields.js
@@ -50,6 +50,15 @@ class ModalSubmitFields {
   getTextInputValue(customId) {
     return this.getField(customId, ComponentType.TextInput).value;
   }
+
+  /**
+   * Gets the values of a select menu component given a custom id
+   * @param {string} customId The custom id for the select menu component
+   * @returns {string[]}
+   */
+  getSelectMenuValues(customId) {
+    return this.getField(customId, ComponentType.SelectMenu).data.values;
+  }
 }
 
 module.exports = ModalSubmitFields;
diff --git a/src/structures/SelectMenuBuilder.js b/src/structures/SelectMenuBuilder.js
index 915f3bc688a758cf60da387ee86862ecc8d66903..3bccdb1d9b756940f2c7f8d532c6f779e44ae0a4 100644
--- a/src/structures/SelectMenuBuilder.js
+++ b/src/structures/SelectMenuBuilder.js
@@ -1,54 +1,12 @@
 'use strict';
 
 const { SelectMenuBuilder: BuildersSelectMenu, isJSONEncodable } = require('@discordjs/builders');
-const Transformers = require('../util/Transformers');
-const Util = require('../util/Util');
 
 /**
  * Class used to build select menu components to be sent through the API
  * @extends {BuildersSelectMenu}
  */
 class SelectMenuBuilder extends BuildersSelectMenu {
-  constructor({ options, ...data } = {}) {
-    super(
-      Transformers.toSnakeCase({
-        ...data,
-        options: options?.map(({ emoji, ...option }) => ({
-          ...option,
-          emoji: emoji && typeof emoji === 'string' ? Util.parseEmoji(emoji) : emoji,
-        })),
-      }),
-    );
-  }
-
-  /**
-   * Adds options to this select menu
-   * @param {APISelectMenuOption[]} options The options to add to this select menu
-   * @returns {SelectMenuBuilder}
-   */
-  addOptions(options) {
-    return super.addOptions(
-      options.map(({ emoji, ...option }) => ({
-        ...option,
-        emoji: emoji && typeof emoji === 'string' ? Util.parseEmoji(emoji) : emoji,
-      })),
-    );
-  }
-
-  /**
-   * Sets the options on this select menu
-   * @param {APISelectMenuOption[]} options The options to set on this select menu
-   * @returns {SelectMenuBuilder}
-   */
-  setOptions(options) {
-    return super.setOptions(
-      options.map(({ emoji, ...option }) => ({
-        ...option,
-        emoji: emoji && typeof emoji === 'string' ? Util.parseEmoji(emoji) : emoji,
-      })),
-    );
-  }
-
   /**
    * Creates a new select menu builder from json data
    * @param {JSONEncodable<APISelectMenuComponent> | APISelectMenuComponent} other The other data
diff --git a/src/structures/SelectMenuOptionBuilder.js b/src/structures/SelectMenuOptionBuilder.js
index 981b8d0ecef4b93d96c798f5ecc51143761d0424..137cc4590ce3ec9240b03db0c2a7bb688d8f778a 100644
--- a/src/structures/SelectMenuOptionBuilder.js
+++ b/src/structures/SelectMenuOptionBuilder.js
@@ -1,34 +1,12 @@
 'use strict';
 
 const { SelectMenuOptionBuilder: BuildersSelectMenuOption, isJSONEncodable } = require('@discordjs/builders');
-const Transformers = require('../util/Transformers');
-const Util = require('../util/Util');
 
 /**
  * Represents a select menu option builder.
  * @extends {BuildersSelectMenuOption}
  */
 class SelectMenuOptionBuilder extends BuildersSelectMenuOption {
-  constructor({ emoji, ...data } = {}) {
-    super(
-      Transformers.toSnakeCase({
-        ...data,
-        emoji: emoji && typeof emoji === 'string' ? Util.parseEmoji(emoji) : emoji,
-      }),
-    );
-  }
-  /**
-   * Sets the emoji to display on this option
-   * @param {ComponentEmojiResolvable} emoji The emoji to display on this option
-   * @returns {SelectMenuOptionBuilder}
-   */
-  setEmoji(emoji) {
-    if (typeof emoji === 'string') {
-      return super.setEmoji(Util.parseEmoji(emoji));
-    }
-    return super.setEmoji(emoji);
-  }
-
   /**
    * Creates a new select menu option builder from JSON data
    * @param {JSONEncodable<APISelectMenuOption>|APISelectMenuOption} other The other data
diff --git a/src/util/Util.js b/src/util/Util.js
index 6b3a693d4c72bcd273d15d5b095b21684b857b3e..4b1630b12354c73b7966a84e53a455258d751a87 100644
--- a/src/util/Util.js
+++ b/src/util/Util.js
@@ -359,17 +359,14 @@ class Util extends null {
    * @param {string} data The string resolvable to resolve
    * @param {Function} [error] The Error constructor to instantiate. Defaults to Error
    * @param {string} [errorMessage] The error message to throw with. Defaults to "Expected string, got <data> instead."
-   * @param {boolean} [allowEmpty=true] Whether an empty string should be allowed
    * @returns {string}
    */
   static verifyString(
     data,
     error = Error,
     errorMessage = `Expected a string, got ${data} instead.`,
-    allowEmpty = true,
   ) {
     if (typeof data !== 'string') throw new error(errorMessage);
-    if (!allowEmpty && data.length === 0) throw new error(errorMessage);
     return data;
   }
 
diff --git a/typings/index.d.ts b/typings/index.d.ts
index 65a790c555886fb3d9ada8d12574ed9d23a170ab..44b8d6d6e740ceeb99f01881adf53b5b8e36ccb7 100644
--- a/typings/index.d.ts
+++ b/typings/index.d.ts
@@ -257,7 +257,7 @@ export class ActionRowBuilder<T extends AnyComponentBuilder = AnyComponentBuilde
 }
 
 export type MessageActionRowComponent = ButtonComponent | SelectMenuComponent;
-export type ModalActionRowComponent = TextInputComponent;
+export type ModalActionRowComponent = TextInputComponent | SelectMenuComponent;
 
 export class ActionRow<T extends MessageActionRowComponent | ModalActionRowComponent> extends Component<
   APIActionRowComponent<APIMessageActionRowComponent | APIModalActionRowComponent>
@@ -393,20 +393,15 @@ export interface InteractionResponseFields<Cached extends CacheType = CacheType>
   replied: boolean;
   webhook: InteractionWebhook;
   reply(options: InteractionReplyOptions & { fetchReply: true }): Promise<Message>;
-  reply(options: string | MessagePayload | InteractionReplyOptions): Promise<void>;
+  reply(
+    options: string | MessagePayload | InteractionReplyOptions,
+  ): Promise<InteractionResponse<BooleanCache<Cached>>>;
   deleteReply(): Promise<void>;
   editReply(options: string | MessagePayload | WebhookEditMessageOptions): Promise<Message>;
   deferReply(options: InteractionDeferReplyOptions & { fetchReply: true }): Promise<Message>;
-  deferReply(options?: InteractionDeferReplyOptions): Promise<void>;
+  deferReply(options?: InteractionDeferReplyOptions): Promise<InteractionResponse<BooleanCache<Cached>>>;
   fetchReply(): Promise<Message>;
   followUp(options: string | MessagePayload | InteractionReplyOptions): Promise<Message>;
-  showModal(
-    modal:
-      | JSONEncodable<APIModalInteractionResponseCallbackData>
-      | ModalComponentData
-      | APIModalInteractionResponseCallbackData,
-  ): Promise<void>;
-  awaitModalSubmit(options: AwaitModalSubmitOptions<ModalSubmitInteraction>): Promise<ModalSubmitInteraction<Cached>>;
 }
 
 export type BooleanCache<T extends CacheType> = T extends 'cached' ? true : false;
@@ -1459,7 +1454,7 @@ export class IntentsBitField extends BitField<GatewayIntentsString> {
   public static resolve(bit?: BitFieldResolvable<GatewayIntentsString, number>): number;
 }
 
-export type CacheType = 'cached' | 'raw' | undefined;
+export type CacheType = 'cached' | 'raw' | 'DM' | 'cachedOrDM' | undefined;
 
 export type CacheTypeReducer<
   State extends CacheType,
@@ -1467,12 +1462,18 @@ export type CacheTypeReducer<
   RawType = CachedType,
   PresentType = CachedType | RawType,
   Fallback = PresentType | null,
+  DMType = CachedType,
+  CachedOrDMType = CachedType | DMType,
 > = [State] extends ['cached']
   ? CachedType
   : [State] extends ['raw']
   ? RawType
   : [State] extends ['raw' | 'cached']
   ? PresentType
+  : [State] extends ['DM']
+  ? DMType
+  : [State] extends ['cachedOrDM']
+  ? CachedOrDMType
   : Fallback;
 
 export type AnyInteraction<Cached extends CacheType = CacheType> =
@@ -1494,7 +1495,8 @@ export class Interaction<Cached extends CacheType = CacheType> extends Base {
     GuildTextBasedChannel | null,
     GuildTextBasedChannel | null,
     GuildTextBasedChannel | null,
-    TextBasedChannel | null
+    TextBasedChannel | null,
+    DMChannel | null
   >;
   public channelId: Snowflake | null;
   public get createdAt(): Date;
@@ -1502,7 +1504,7 @@ export class Interaction<Cached extends CacheType = CacheType> extends Base {
   public get guild(): CacheTypeReducer<Cached, Guild, null>;
   public guildId: CacheTypeReducer<Cached, Snowflake>;
   public id: Snowflake;
-  public member: CacheTypeReducer<Cached, GuildMember, APIInteractionGuildMember>;
+  public member: CacheTypeReducer<Cached, GuildMember, APIInteractionGuildMember, GuildMember | APIInteractionGuildMember | null, null>;
   public readonly token: string;
   public type: InteractionType;
   public user: User;
@@ -1719,16 +1721,15 @@ export class Message<Cached extends boolean = boolean> extends Base {
 }
 
 export class AttachmentBuilder {
-  public constructor(attachment: BufferResolvable | Stream, data?: AttachmentData);
+  public constructor(data?: Partial<AttachmentPayload>);
   public attachment: BufferResolvable | Stream;
-  public description: string | null;
-  public name: string | null;
+  public data: Exclude<AttachmentPayload, 'attachment'>;
   public get spoiler(): boolean;
   public setDescription(description: string): this;
-  public setFile(attachment: BufferResolvable | Stream, name?: string): this;
+  public setFile(attachment: BufferResolvable | Stream): this;
   public setName(name: string): this;
   public setSpoiler(spoiler?: boolean): this;
-  public toJSON(): unknown;
+  public toJSON(): AttachmentPayload;
   public static from(other: JSONEncodable<AttachmentPayload>): AttachmentBuilder;
 }
 
@@ -1934,6 +1935,7 @@ export class ModalSubmitFields {
   public getField<T extends ComponentType>(customId: string, type: T): { type: T } & ModalData;
   public getField(customId: string, type?: ComponentType): ModalData;
   public getTextInputValue(customId: string): string;
+  public getSelectMenuValues(customId: string): string[];
 }
 
 export interface ModalMessageModalSubmitInteraction<Cached extends CacheType = CacheType>
@@ -3532,11 +3534,6 @@ export interface BaseApplicationCommandData {
   defaultPermission?: boolean;
 }
 
-export interface AttachmentData {
-  name?: string;
-  description?: string;
-}
-
 export type CommandOptionDataTypeResolvable = ApplicationCommandOptionType;
 
 export type CommandOptionChannelResolvableType = ApplicationCommandOptionType.Channel;
@@ -4766,10 +4763,8 @@ export interface MessageEditOptions {
   files?: (
     | BufferResolvable
     | Stream
-    | JSONEncodable<APIAttachment>
-    | Attachment
-    | AttachmentBuilder
     | AttachmentPayload
+    | JSONEncodable<AttachmentPayload>
   )[];
   flags?: BitFieldResolvable<MessageFlagsString, number>;
   allowedMentions?: MessageMentionOptions;
@@ -4824,10 +4819,8 @@ export interface MessageOptions {
   files?: (
     | BufferResolvable
     | Stream
-    | JSONEncodable<APIAttachment>
-    | Attachment
-    | AttachmentBuilder
     | AttachmentPayload
+    | JSONEncodable<AttachmentPayload>
   )[];
   reply?: ReplyOptions;
   stickers?: StickerResolvable[];
