diff --git a/lib/fetch/request.js b/lib/fetch/request.js
index 080a5d7bfa37d3851a11b915467d68146bd63bb2..058210ac65f44c4170e363c39d6a9c49a7462164 100644
--- a/lib/fetch/request.js
+++ b/lib/fetch/request.js
@@ -30,6 +30,7 @@ const { URLSerializer } = require('./dataURL')
 const { kHeadersList } = require('../core/symbols')
 const assert = require('assert')
 const { setMaxListeners, getEventListeners, defaultMaxListeners } = require('events')
+const { aborted } = require('util');
 
 let TransformStream = globalThis.TransformStream
 
@@ -354,10 +355,6 @@ class Request {
       if (signal.aborted) {
         ac.abort(signal.reason)
       } else {
-        const abort = function () {
-          ac.abort(this.reason)
-        }
-
         // Third-party AbortControllers may not work with these.
         // See https://github.com/nodejs/undici/pull/1910#issuecomment-1464495619
         try {
@@ -366,8 +363,7 @@ class Request {
           }
         } catch {}
 
-        signal.addEventListener('abort', abort, { once: true })
-        requestFinalizer.register(this, { signal, abort })
+        aborted(signal, this).then(() => ac.abort(signal.reason))
       }
     }
 
