diff --git a/lib/api/api-request.js b/lib/api/api-request.js
index b4674878d2e9d77e168c54a6b0fdaf482d4c05a7..a6edb8b227b7255355474ac4c281beec85af9de9 100644
--- a/lib/api/api-request.js
+++ b/lib/api/api-request.js
@@ -101,6 +101,7 @@ class RequestHandler extends AsyncResource {
 
       this.runInAsyncScope(callback, null, null, {
         statusCode,
+	    statusMessage,
         headers,
         trailers: this.trailers,
         opaque,
diff --git a/types/dispatcher.d.ts b/types/dispatcher.d.ts
index 8744f04e2d7083a4639da974311bf71d414fc2d4..e6970479e820f540a0654be5dc3c27b9d5d88b5a 100644
--- a/types/dispatcher.d.ts
+++ b/types/dispatcher.d.ts
@@ -41,8 +41,8 @@ declare class Dispatcher extends EventEmitter {
 declare namespace Dispatcher {
   export interface DispatchOptions {
     origin?: string | URL;
-    path: string;
-    method: HttpMethod;
+    path?: string;
+    method?: HttpMethod;
     /** Default: `null` */
     body?: string | Buffer | Uint8Array | Readable | null | FormData;
     /** Default: `null` */
@@ -112,6 +112,7 @@ declare namespace Dispatcher {
   }
   export interface ResponseData {
     statusCode: number;
+    statusMessage: string;
     headers: IncomingHttpHeaders;
     body: BodyReadable & BodyMixin;
     trailers: Record<string, string>;
