diff --git a/src/client.js b/src/client.js
index 6f1182337db80b241aee9b195cd522e788d252e8..8ceac8e0bef51a59af686f1a9617cb69128efec9 100644
--- a/src/client.js
+++ b/src/client.js
@@ -14,7 +14,7 @@ const closeTimeout = 30 * 1000
 
 class Client extends EventEmitter {
   constructor (isServer, version, customPackets, hideErrors = false) {
-    super()
+    super({ captureRejections: true })
     this.customPackets = customPackets
     this.version = version
     this.isServer = !!isServer
